modules/imgcodecs/src/grfmt_pfm.cpp

```diff
@@ -0,0 +1,314 @@
+  uint16_t endianness = 0xbeef;
+  unsigned char * tmp = (unsigned char *)&endianness;
+  return *tmp == 0xef;
+}
+
+bool is_byte_order_swapped(float scale)
+{
+  // ".pfm" format file specifies that:
+  // positive scale means big endianess;
+  // negative scale means little endianess.
+
+  if (is_little_endian()) {
+    return scale >= 0.0;
+  } else {
+    return scale < 0.0;
+  }
+}
+
+void swap_endianess(float& f)
+{
+  static const uint32_t A(0x000000ffU);
+  static const uint32_t B(0x0000ff00U);
+  static const uint32_t C(0x00ff0000U);
+  static const uint32_t D(0xff000000U);
+
+  union {
+    uint32_t i;
+    float f;
+  } cast;
+
+  cast.f = f;
+  cast.i = ( (cast.i & A) << 24 )
+         | ( (cast.i & B) <<  8 )
+         | ( (cast.i & C) >>  8 )
+         | ( (cast.i & D) >> 24 );
+  f = cast.f;
+}
+
+template<typename T> T atoT(const std::string& s);
+template<> int atoT<int>(const std::string& s) { return std::atoi(s.c_str()); }
+template<> float atoT<float>(const std::string& s) { return std::atof(s.c_str()); }
+
+template<typename T>
+T read_number(cv::RLByteStream& strm)
+{
+  // should be enogh to take string representation of any number
+  const size_t buffer_size = 2048;
+
+  std::vector<char> buffer(buffer_size, 0);
+  for (size_t i = 0; i < buffer_size; ++i) {
+    char c = strm.getByte();
+    if (std::isspace(c)) {
+      break;
+    }
+    buffer[i] = c;
+  }
+  const std::string str(buffer.begin(), buffer.end());
+  return atoT<T>(str);
+}
+
+template<typename T> void write_anything(cv::WLByteStream& strm, const T& t)
+{
+  const std::string str = std::to_string(t);
+  strm.putBytes(str.c_str(), str.size());
+}
+
+}
+
+namespace cv {
+
+PFMDecoder::~PFMDecoder()
+{
+}
+
+PFMDecoder::PFMDecoder()
+{
+  m_strm.close();
+}
+
+bool PFMDecoder::readHeader()
+{
+  if (m_buf.empty()) {
+    if (!m_strm.open(m_filename)) {
+      return false;
+    }
+  } else {
+    if (!m_strm.open(m_buf)) {
+      return false;
+    }
+  }
+
+  if (m_strm.getByte() != 'P') {
+    CV_Error(Error::StsError, "Unexpected file type (expected P)");
+    return false;
+  }
+
+  switch (m_strm.getByte()) {
+  case 'f':
+    m_channels = 1;
+    break;
+  case 'F':
+    m_channels = 3;
+    break;
+  default:
+    CV_Error(Error::StsError, "Unexpected file type (expected `f` or `F`)");
+    return false;
+  }
+
+  if ('\n' != m_strm.getByte()) {
+    CV_Error(Error::StsError, "Unexpected header format (expected line break)");
+    return false;
+  }
+
+
+  m_width = read_number<int>(m_strm);
+  m_height = read_number<int>(m_strm);
+  m_scale_factor = read_number<float>(m_strm);
+  m_swap_byte_order = is_byte_order_swapped(m_scale_factor);
+
+  return true;
+}
+
+bool PFMDecoder::readData(Mat& mat)
+{
+  if (!m_strm.isOpened()) {
+    CV_Error(Error::StsError, "Unexpected status in data stream");
+    return false;
+  }
+
+  if (m_channels == 3) {
+    mat = Mat3f(m_height, m_width);
+  } else if (m_channels == 1) {
+    mat = Mat1f(m_height, m_width);
```

Comment:
`readData()` method should not re-create `Mat`.
You need to check/fit input data into passed Mat.

`m_type` should be properly filled.
or implement `.type()`method.

Also all used `imread()` calls (in tests) should be called with `IMREAD_UNCHANGED` flag (or `CV_LOAD_IMAGE_ANYDEPTH | CV_LOAD_IMAGE_ANYCOLOR`).

Reply:
So if `IMREAD_UNCHANGED` is not passed to `imread`, the type changes silently to `CV_8UC1` or `CV_8UC3`, right?
But this will not work, as the image buffer will be too small (assuming 1 byte elements rather than 4 bytes). Hence it will crash or at least fail to read the image.
Is that what we want? Shouldn't no flags mean to load it as CV_32F1 or CV_32F3 in the PFM case?

Reply:
> the type changes silently to CV_8UC1 or CV_8UC3

Right (8UC3 - if `IMREAD_COLOR` flag passed). This is default behavior of `cv::imread()` (independent of any used codec).

You should read into your buffer / Mat. And after reading, convert to passed `img`.

Or leave an exception. There is no convenient way to convert float number into int8.